### ISSUE TITLE
Make P4Tools' reachability analysis more general and reuse existing compiler code

### ISSUE THREAD

**Issue by fruffy** [p4lang/p4c#3581]
Emerged from the discussion in #3569.

**Comment by fruffy** [comment]
@mbudiu-vmw mentioned that there are a few existing facilities that perform similar analysis in p4c.  We can take a closer look at how to improve the existing pass before migrating it to the mid end.  I do not quite know yet, whether we can reuse existing code. This issue is intended to continue the discussion from the merged PR and to track this issue.

**Comment by VolodymyrPeschanenkoLitSoft** [comment]
Let's start with the description of the task which was solved by the reachability feature.  For each pair `(s, e)` of `IR::Node`'s it should return true if `e` is reachable from `s` by control flow. Otherwise, It should return `false`. Considered nodes should contain IR::Anontation name which could be used from the user's input to get the corresponding IR::Node which represents it or the parent node which contains it.
Let's consider different possibilities for the implementation of this feature. 
1. [graphs backend](https://github.com/p4lang/p4c/tree/main/backends/graphs). It is placed in `backends/graphs` folder. It was implemented based on the`boost::subgraph` class. It doesn't store all nodes in that subgraph. For example, [Here](https://github.com/p4lang/p4c/blob/19fe76e943080fd57ffe8b97cb8b8a9b743f8b3e/backends/graphs/controls.cpp#L233) it works with `IR::Assigment` statement. It stored it [here](https://github.com/p4lang/p4c/blob/19fe76e943080fd57ffe8b97cb8b8a9b743f8b3e/backends/graphs/graphs.h#L219) for the corresponding vertex. In this case, all assignments will be out of `boost::subgraph `searching procedure and it'll hard to find the corresponding subgraph for such assignments.
Analysis result: these classes can't be used for reachability feature implementation, but in general we can consider changing the main class for implementation from [CallGraph](https://github.com/p4lang/p4c/blob/19fe76e943080fd57ffe8b97cb8b8a9b743f8b3e/frontends/p4/callGraph.h#L40) to the `boost::subgraph`. However, I think that in that case, the class [P4ProgramDCGCreator](https://github.com/p4lang/p4c/blob/19fe76e943080fd57ffe8b97cb8b8a9b743f8b3e/backends/p4tools/common/compiler/reachability.h#L88) will be almost the same.

2. [ControlFlowVisitor](https://github.com/p4lang/p4c/blob/19fe76e943080fd57ffe8b97cb8b8a9b743f8b3e/ir/visitor.h#L372).  The [P4ProgramDCGCreator](https://github.com/p4lang/p4c/blob/19fe76e943080fd57ffe8b97cb8b8a9b743f8b3e/backends/p4tools/common/compiler/reachability.h#L88) doesn't use copy object before applying to the arguments of the nodes. It uses the `visit` function because the visitor crashed in the opposite case.
Analysis result: The [P4ProgramDCGCreator](https://github.com/p4lang/p4c/blob/19fe76e943080fd57ffe8b97cb8b8a9b743f8b3e/backends/p4tools/common/compiler/reachability.h#L88) can use [ControlFlowVisitor](https://github.com/p4lang/p4c/blob/19fe76e943080fd57ffe8b97cb8b8a9b743f8b3e/ir/visitor.h#L372) is it make a copy of an object before applying to the arguments for joins detection.

3. def-use and CFG in a process.

**Comment by mbudiu-vmw** [comment]
Can you define more precisely what "reachable" means? Are you talking about statements? Expressions? There are many kinds of IR nodes. If you consider procedures or actions, do you want an inter-procedural or intra-procedural analysis? If the former, do you want it context-sensitive or context-insensitive? Our visitors are not really designed for inter-procedural analyses, but they can be used by creating new visitors when you visit a procedure, we have many instances, e.g., in def-use.

I don't understand the problem about the crash. The Visitors are 100% safe if used correctly.

**Comment by fruffy** [comment]
The original goal of this reachability analysis is to build a map for each statement in the program which contains a list of accessible downstream nodes (for now, annotations, tables, parser states. I believe this list can be expanded). The analysis should be both inter-procedural and context-insensitive. 

Concretely, the long-term goal here is to allow tools such as P4Testgen to prune generated tests based on input queries. For example, one may only want to generate tests that hit the IPv4 parser state. All other tests should be eliminated as early as possible. The reachability analysis is intended to be a building block of this feature. Its design must reflect that. 

>I don't understand the problem about the crash. The Visitors are 100% safe if used correctly.

If I understand this correctly, using `apply` within a preorder/postorder function using the same visitor leads to a crash. The only way to solve this is to create a new visitor, copy some of the state, then apply. I am not sure whether this is the case for the ControlFlowVisitor. @VolodymyrPeschanenkoLitSoft can clarify.

**Comment by VolodymyrPeschanenkoLitSoft** [comment]
> I don't understand the problem about the crash. The Visitors are 100% safe if used correctly.

Here is a simple example:

```
class NewInspector : public Inspector {
  public:
    NewInspector() {};
    bool preorder(const IR::P4Control* control) override {
        control->body->apply(*this);
        return false;
    }
};
...
 NewInspector insp;
 program->apply(insp);
```
Here `program` is a IR::P4Program. I tested it on `[this](https://github.com/p4lang/p4c/blob/main/testdata/p4_16_samples/stack_complex-bmv2.p4)` example and it crashed [here](https://github.com/p4lang/p4c/blob/7aaeddde55f50f31c10234b9691e98011e72da97/ir/visitor.cpp#L328) (visited was a nullptr).

**Comment by VolodymyrPeschanenkoLitSoft** [comment]
3. [def-use](https://github.com/p4lang/p4c/blob/7aaeddde55f50f31c10234b9691e98011e72da97/frontends/p4/simplifyDefUse.h#L26) can't be used for reacchability feature, because it doesn't store information about transitions.
4. [CFG](https://github.com/p4lang/p4c/blob/main/backends/bmv2/common/controlFlowGraph.h). It is in backends/bmv2/common folder. It doesn't support IR::P4Parser, IR::ParserState, IR::Annotation. In general, we can use it for implemeentation of the reachability feature, but it should be expanded by the additional makeNode functions [here](https://github.com/p4lang/p4c/blob/7aaeddde55f50f31c10234b9691e98011e72da97/backends/bmv2/common/controlFlowGraph.h#L153) and by new functions to the Inspector [here](https://github.com/p4lang/p4c/blob/7aaeddde55f50f31c10234b9691e98011e72da97/backends/bmv2/common/controlFlowGraph.cpp#L185). It doesn't have function which can define a place in `CFG` from `IR::Node`.

As a result I can say, that CallGraph was implemented with help of oredered_maps [here](https://github.com/p4lang/p4c/blob/7aaeddde55f50f31c10234b9691e98011e72da97/frontends/p4/callGraph.h#L44). It is easy to get some state inside CFG by IR::Node and the main data structure for implementing of the reachability feature was selected correctly. 

The [CFGBuilder](https://github.com/p4lang/p4c/blob/7aaeddde55f50f31c10234b9691e98011e72da97/backends/bmv2/common/controlFlowGraph.cpp#L263) and [P4ProgramDCGCreator](https://github.com/p4lang/p4c/blob/7aaeddde55f50f31c10234b9691e98011e72da97/backends/p4tools/common/compiler/reachability.cpp#L301) classes use `visited` function for sub-path traversal which makes unusable the [ControlFlowVisitor](https://github.com/p4lang/p4c/blob/7aaeddde55f50f31c10234b9691e98011e72da97/ir/visitor.h#L372) class for the reachability feature because its function [filter_join_point](https://github.com/p4lang/p4c/blob/7aaeddde55f50f31c10234b9691e98011e72da97/ir/visitor.h#L395) will not be called.

@fruffy, @mbudiu-vmw What do you think?

**Comment by mbudiu-vmw** [comment]
Yes, you should not reapply the same visitor to a node that is already being visited.
